// motia/dist/index.mjs
import { Logger, getContext, getContext as getContext$1, init } from "iii-sdk";
import "esbuild";
import { existsSync } from "fs";
import { globSync } from "glob";
import path from "path";
import { v5 } from "uuid";
import fs from "fs/promises";
import { SpanStatusCode, withSpan } from "iii-sdk/telemetry";
import * as z from "zod";

//#region src/guards.ts
const isApiTrigger = (trigger) => trigger.type === "http";
const isQueueTrigger = (trigger) => trigger.type === "queue";
const isCronTrigger = (trigger) => trigger.type === "cron";
const isStateTrigger = (trigger) => trigger.type === "state";
const isStreamTrigger = (trigger) => trigger.type === "stream";
const getApiTriggers = (step$1) => step$1.config.triggers.filter(isApiTrigger);
const getQueueTriggers = (step$1) => step$1.config.triggers.filter(isQueueTrigger);
const getCronTriggers = (step$1) => step$1.config.triggers.filter(isCronTrigger);

//#endregion
//#region src/multi-trigger.ts
function multiTriggerStep(config) {
	const collectedHandlers = {};
	const createUnifiedHandler = () => {
		return (async (input, ctx) => {
			if (ctx.trigger.type === "queue" && collectedHandlers.queue) return collectedHandlers.queue(input, ctx);
			if (ctx.trigger.type === "http" && collectedHandlers.http) return collectedHandlers.http(input, ctx);
			if (ctx.trigger.type === "cron" && collectedHandlers.cron) return collectedHandlers.cron(ctx);
			if (ctx.trigger.type === "state" && collectedHandlers.state) return collectedHandlers.state(input, ctx);
			if (ctx.trigger.type === "stream" && collectedHandlers.stream) return collectedHandlers.stream(input, ctx);
			ctx.logger.warn(`No handler defined for trigger type: ${ctx.trigger.type}`, {
				availableHandlers: Object.keys(collectedHandlers),
				triggerType: ctx.trigger.type
			});
			throw new Error(`No handler defined for trigger type: ${ctx.trigger.type}. Available handlers: ${Object.keys(collectedHandlers).join(", ")}`);
		});
	};
	const builder = {
		config,
		onQueue(handler) {
			collectedHandlers.queue = handler;
			return {
				...builder,
				handlers: () => ({
					config,
					handler: createUnifiedHandler()
				})
			};
		},
		onHttp(handler) {
			collectedHandlers.http = handler;
			return {
				...builder,
				handlers: () => ({
					config,
					handler: createUnifiedHandler()
				})
			};
		},
		onCron(handler) {
			collectedHandlers.cron = handler;
			return {
				...builder,
				handlers: () => ({
					config,
					handler: createUnifiedHandler()
				})
			};
		},
		onState(handler) {
			collectedHandlers.state = handler;
			return {
				...builder,
				handlers: () => ({
					config,
					handler: createUnifiedHandler()
				})
			};
		},
		onStream(handler) {
			collectedHandlers.stream = handler;
			return {
				...builder,
				handlers: () => ({
					config,
					handler: createUnifiedHandler()
				})
			};
		},
		handlers(h) {
			if (h) Object.assign(collectedHandlers, h);
			return {
				config,
				handler: createUnifiedHandler()
			};
		}
	};
	return builder;
}

//#endregion
//#region src/new/build/loader.ts
const STEP_NAMESPACE = "7f1c3ff2-9b00-4d0a-bdd7-efb8bca49d4f";
const generateStepId = (filePath) => {
	return v5(filePath, STEP_NAMESPACE);
};

//#endregion
//#region src/new/iii.ts
const engineWsUrl = process.env.III_URL ?? "ws://localhost:49134";
const createIII = (otelConfig) => {
	return init(engineWsUrl, { otel: {
		enabled: true,
		serviceName: "motia",
		...otelConfig
	} });
};
let instance;
const getInstance = () => {
	if (!instance) instance = createIII();
	return instance;
};
const initIII = (otelConfig) => {
	instance = createIII(otelConfig);
	return instance;
};

//#endregion
//#region src/new/build/generate-index.ts
const toRelativePosix = (file) => {
	return "./" + path.relative(process.cwd(), file).replace(/\\/g, "/");
};
const getStepFilesFromDir = (dir) => {
	if (!existsSync(dir)) return [];
	return globSync("**/*.step.{ts,js}", {
		absolute: true,
		cwd: dir
	}).map(toRelativePosix);
};

//#endregion
//#region src/new/setup-step-endpoint.ts
const getFeatures = async (filePath) => {
	const stat = await fs.stat(`${filePath}-features.json`).catch(() => null);
	if (!stat || stat.isDirectory()) return [];
	try {
		const content = await fs.readFile(`${filePath}-features.json`, "utf8");
		return JSON.parse(content);
	} catch {
		return [];
	}
};
function setupStepEndpoint(iii) {
	const function_id = "motia_step_get";
	iii.registerFunction({ id: function_id }, async (req) => {
		const id = req.path_params.stepId;
		const step$1 = [...getStepFilesFromDir(path.join(process.cwd(), "steps")), ...getStepFilesFromDir(path.join(process.cwd(), "src"))].find((step$2) => generateStepId(step$2) === id);
		if (!step$1) return {
			status_code: 404,
			body: { error: "Step not found" }
		};
		try {
			return {
				status_code: 200,
				body: {
					id,
					content: await fs.readFile(step$1, "utf8"),
					features: await getFeatures(step$1.replace(`${path.sep}src${path.sep}`, `${path.sep}tutorial${path.sep}`))
				}
			};
		} catch (error) {
			console.error("Error reading step file:", error);
			return {
				status_code: 500,
				body: { error: "Failed to read step file" }
			};
		}
	});
	iii.registerTrigger({
		trigger_type: "http",
		function_id,
		config: {
			api_path: "__motia/step/:stepId",
			http_method: "GET",
			description: "Get a step"
		}
	});
}

//#endregion
//#region src/new/state.ts
var StateManager = class {
	async get(scope, key) {
		return withSpan("state::get", {}, async (span) => {
			span.setAttribute("motia.state.scope", scope);
			span.setAttribute("motia.state.key", key);
			try {
				return await getInstance().call("state::get", {
					scope,
					key
				});
			} catch (err) {
				span.setStatus({
					code: SpanStatusCode.ERROR,
					message: String(err)
				});
				span.recordException(err);
				throw err;
			}
		});
	}
	async set(scope, key, data) {
		return withSpan("state::set", {}, async (span) => {
			span.setAttribute("motia.state.scope", scope);
			span.setAttribute("motia.state.key", key);
			try {
				return await getInstance().call("state::set", {
					scope,
					key,
					data
				});
			} catch (err) {
				span.setStatus({
					code: SpanStatusCode.ERROR,
					message: String(err)
				});
				span.recordException(err);
				throw err;
			}
		});
	}
	async delete(scope, key) {
		return withSpan("state::delete", {}, async (span) => {
			span.setAttribute("motia.state.scope", scope);
			span.setAttribute("motia.state.key", key);
			try {
				return await getInstance().call("state::delete", {
					scope,
					key
				});
			} catch (err) {
				span.setStatus({
					code: SpanStatusCode.ERROR,
					message: String(err)
				});
				span.recordException(err);
				throw err;
			}
		});
	}
	async list(scope) {
		return withSpan("state::list", {}, async (span) => {
			span.setAttribute("motia.state.scope", scope);
			try {
				return await getInstance().call("state::list", { scope });
			} catch (err) {
				span.setStatus({
					code: SpanStatusCode.ERROR,
					message: String(err)
				});
				span.recordException(err);
				throw err;
			}
		});
	}
	async listGroups() {
		return withSpan("state::list_groups", {}, async (span) => {
			try {
				return await getInstance().call("state::list_groups", {});
			} catch (err) {
				span.setStatus({
					code: SpanStatusCode.ERROR,
					message: String(err)
				});
				span.recordException(err);
				throw err;
			}
		});
	}
	async update(scope, key, ops) {
		return withSpan("state::update", {}, async (span) => {
			span.setAttribute("motia.state.scope", scope);
			span.setAttribute("motia.state.key", key);
			try {
				return await getInstance().call("state::update", {
					scope,
					key,
					ops
				});
			} catch (err) {
				span.setStatus({
					code: SpanStatusCode.ERROR,
					message: String(err)
				});
				span.recordException(err);
				throw err;
			}
		});
	}
	async clear(scope) {
		return withSpan("state::clear", {}, async (span) => {
			span.setAttribute("motia.state.scope", scope);
			try {
				const items = await this.list(scope);
				for (const item of items) await this.delete(scope, item.id);
			} catch (err) {
				span.setStatus({
					code: SpanStatusCode.ERROR,
					message: String(err)
				});
				span.recordException(err);
				throw err;
			}
		});
	}
};
const stateManager = new StateManager();

//#endregion
//#region src/new/stream.ts
var Stream = class {
	constructor(config) {
		this.config = config;
	}
	async get(groupId, itemId) {
		return withSpan("stream::get", {}, async (span) => {
			span.setAttribute("motia.stream.name", this.config.name);
			span.setAttribute("motia.stream.group_id", groupId);
			span.setAttribute("motia.stream.item_id", itemId);
			try {
				return await getInstance().call("stream::get", {
					stream_name: this.config.name,
					group_id: groupId,
					item_id: itemId
				});
			} catch (err) {
				span.setStatus({
					code: SpanStatusCode.ERROR,
					message: String(err)
				});
				span.recordException(err);
				throw err;
			}
		});
	}
	async set(groupId, itemId, data) {
		return withSpan("stream::set", {}, async (span) => {
			span.setAttribute("motia.stream.name", this.config.name);
			span.setAttribute("motia.stream.group_id", groupId);
			span.setAttribute("motia.stream.item_id", itemId);
			try {
				return await getInstance().call("stream::set", {
					stream_name: this.config.name,
					group_id: groupId,
					item_id: itemId,
					data
				});
			} catch (err) {
				span.setStatus({
					code: SpanStatusCode.ERROR,
					message: String(err)
				});
				span.recordException(err);
				throw err;
			}
		});
	}
	async delete(groupId, itemId) {
		return withSpan("stream::delete", {}, async (span) => {
			span.setAttribute("motia.stream.name", this.config.name);
			span.setAttribute("motia.stream.group_id", groupId);
			span.setAttribute("motia.stream.item_id", itemId);
			try {
				return await getInstance().call("stream::delete", {
					stream_name: this.config.name,
					group_id: groupId,
					item_id: itemId
				});
			} catch (err) {
				span.setStatus({
					code: SpanStatusCode.ERROR,
					message: String(err)
				});
				span.recordException(err);
				throw err;
			}
		});
	}
	async list(groupId) {
		return withSpan("stream::list", {}, async (span) => {
			span.setAttribute("motia.stream.name", this.config.name);
			span.setAttribute("motia.stream.group_id", groupId);
			try {
				return await getInstance().call("stream::list", {
					stream_name: this.config.name,
					group_id: groupId
				});
			} catch (err) {
				span.setStatus({
					code: SpanStatusCode.ERROR,
					message: String(err)
				});
				span.recordException(err);
				throw err;
			}
		});
	}
	async update(groupId, itemId, ops) {
		return withSpan("stream::update", {}, async (span) => {
			span.setAttribute("motia.stream.name", this.config.name);
			span.setAttribute("motia.stream.group_id", groupId);
			span.setAttribute("motia.stream.item_id", itemId);
			try {
				return await getInstance().call("stream::update", {
					stream_name: this.config.name,
					group_id: groupId,
					item_id: itemId,
					ops
				});
			} catch (err) {
				span.setStatus({
					code: SpanStatusCode.ERROR,
					message: String(err)
				});
				span.recordException(err);
				throw err;
			}
		});
	}
	async listGroups() {
		return withSpan("stream::list_groups", {}, async (span) => {
			span.setAttribute("motia.stream.name", this.config.name);
			try {
				return await getInstance().call("stream::list_groups", { stream_name: this.config.name });
			} catch (err) {
				span.setStatus({
					code: SpanStatusCode.ERROR,
					message: String(err)
				});
				span.recordException(err);
				throw err;
			}
		});
	}
	async send(channel, event) {
		return withSpan("stream::send", {}, async (span) => {
			span.setAttribute("motia.stream.name", this.config.name);
			span.setAttribute("motia.stream.group_id", channel.groupId);
			if (channel.id) span.setAttribute("motia.stream.item_id", channel.id);
			try {
				await getInstance().call("stream::send", {
					stream_name: this.config.name,
					group_id: channel.groupId,
					id: channel.id,
					type: event.type,
					data: event.data
				});
			} catch (err) {
				span.setStatus({
					code: SpanStatusCode.ERROR,
					message: String(err)
				});
				span.recordException(err);
				throw err;
			}
		});
	}
};

//#endregion
//#region src/new/build/utils.ts
const composeMiddleware = (...middlewares) => {
	return async (req, ctx, handler) => {
		return middlewares.reduceRight((nextHandler, middleware) => () => middleware(req, ctx, nextHandler), handler)();
	};
};
const flowContext = (streamManager, trigger, input) => {
	const { logger, trace } = getContext$1();
	const traceId = trace?.spanContext().traceId ?? crypto.randomUUID();
	const enqueue = async (queue$1) => getInstance().call("enqueue", queue$1);
	return {
		enqueue,
		traceId,
		state: stateManager,
		logger,
		streams: streamManager.streams,
		trigger,
		is: {
			queue: (inp) => trigger.type === "queue",
			http: (inp) => trigger.type === "http",
			cron: (inp) => trigger.type === "cron",
			state: (inp) => trigger.type === "state",
			stream: (inp) => trigger.type === "stream"
		},
		getData: () => {
			if (trigger.type === "http") return input.body;
			return input;
		},
		match: async (handlers) => {
			if (trigger.type === "queue" && handlers.queue) return await handlers.queue(input);
			if (trigger.type === "http" && handlers.http) return await handlers.http(input);
			if (trigger.type === "cron" && handlers.cron) return await handlers.cron();
			if (trigger.type === "state" && handlers.state) return await handlers.state(input);
			if (trigger.type === "stream" && handlers.stream) return await handlers.stream(input);
			if (handlers.default) return await handlers.default(input);
			logger.warn(`No handler matched for trigger type: ${trigger.type}`, {
				availableHandlers: Object.keys(handlers).filter((k) => k !== "default"),
				triggerType: trigger.type
			});
			throw new Error(`No handler matched for trigger type: ${trigger.type}. Available handlers: ${Object.keys(handlers).join(", ")}`);
		}
	};
};
var Motia = class {
	constructor() {
		this.streams = {};
	}
	addStep(config, stepPath, handler, filePath) {
		const step$1 = {
			config,
			handler,
			filePath: stepPath
		};
		const metadata = {
			...step$1.config,
			filePath
		};
		step$1.config.triggers.forEach((trigger, index) => {
			const function_id = `steps::${step$1.config.name}::trigger::${index}`;
			if (isApiTrigger(trigger)) {
				getInstance().registerFunction({
					id: function_id,
					metadata
				}, async (req) => {
					const triggerInfo = {
						type: "http",
						index
					};
					const motiaRequest = {
						pathParams: req.path_params || {},
						queryParams: req.query_params || {},
						body: req.body,
						headers: req.headers || {}
					};
					const context = flowContext(this, triggerInfo, motiaRequest);
					const composedMiddleware = composeMiddleware(...Array.isArray(trigger?.middleware) ? trigger.middleware : []);
					const handlerFn = async () => {
						return await step$1.handler(motiaRequest, context) || {
							status: 200,
							body: null
						};
					};
					const response = await composedMiddleware(motiaRequest, context, handlerFn);
					return {
						status_code: response.status,
						headers: response.headers,
						body: response.body
					};
				});
				const triggerConfig = {
					api_path: trigger.path.startsWith("/") ? trigger.path.substring(1) : trigger.path,
					http_method: trigger.method,
					metadata
				};
				if (trigger.condition) {
					const conditionPath = `${function_id}::conditions::${index}`;
					getInstance().registerFunction({ id: conditionPath }, async (req) => {
						const triggerInfo = {
							type: "http",
							index
						};
						const motiaRequest = {
							pathParams: req.path_params || {},
							queryParams: req.query_params || {},
							body: req.body,
							headers: req.headers || {}
						};
						return trigger.condition?.(motiaRequest, flowContext(this, triggerInfo, motiaRequest));
					});
					triggerConfig._condition_path = conditionPath;
				}
				getInstance().registerTrigger({
					trigger_type: "http",
					function_id,
					config: triggerConfig
				});
			} else if (isQueueTrigger(trigger)) {
				getInstance().registerFunction({
					id: function_id,
					metadata
				}, async (req) => {
					const context = flowContext(this, {
						type: "queue",
						index
					}, req);
					return step$1.handler(req, context);
				});
				const triggerConfig = {
					topic: trigger.topic,
					metadata
				};
				if (trigger.condition) {
					const conditionPath = `${function_id}::conditions::${index}`;
					getInstance().registerFunction({ id: conditionPath }, async (input) => {
						const triggerInfo = {
							type: "queue",
							index
						};
						return trigger.condition?.(input, flowContext(this, triggerInfo, input));
					});
					triggerConfig._condition_path = conditionPath;
				}
				getInstance().registerTrigger({
					type: "queue",
					function_id,
					config: triggerConfig
				});
			} else if (isCronTrigger(trigger)) {
				getInstance().registerFunction({
					id: function_id,
					metadata
				}, async (_req) => {
					const triggerInfo = {
						type: "cron",
						index
					};
					return step$1.handler(void 0, flowContext(this, triggerInfo));
				});
				const triggerConfig = {
					expression: trigger.expression,
					metadata
				};
				if (trigger.condition) {
					const conditionPath = `${function_id}::conditions::${index}`;
					getInstance().registerFunction({ id: conditionPath }, async () => {
						const triggerInfo = {
							type: "cron",
							index
						};
						return trigger.condition?.(void 0, flowContext(this, triggerInfo));
					});
					triggerConfig._condition_path = conditionPath;
				}
				getInstance().registerTrigger({
					type: "cron",
					function_id,
					config: triggerConfig
				});
			} else if (isStateTrigger(trigger)) {
				getInstance().registerFunction({
					id: function_id,
					metadata
				}, async (req) => {
					const context = flowContext(this, {
						type: "state",
						index
					}, req);
					return step$1.handler(req, context);
				});
				const triggerConfig = { metadata };
				if (trigger.condition) {
					const conditionPath = `${function_id}::conditions::${index}`;
					getInstance().registerFunction({ id: conditionPath }, async (input) => {
						const triggerInfo = {
							type: "state",
							index
						};
						return trigger.condition?.(input, flowContext(this, triggerInfo, input));
					});
					triggerConfig.condition_function_id = conditionPath;
				}
				getInstance().registerTrigger({
					type: "state",
					function_id,
					config: triggerConfig
				});
			} else if (isStreamTrigger(trigger)) {
				getInstance().registerFunction({
					id: function_id,
					metadata
				}, async (req) => {
					const context = flowContext(this, {
						type: "stream",
						index
					}, req);
					return step$1.handler(req, context);
				});
				const triggerConfig = {
					metadata,
					stream_name: trigger.streamName,
					group_id: trigger.groupId,
					item_id: trigger.itemId
				};
				if (trigger.condition) {
					const conditionPath = `${function_id}::conditions::${index}`;
					getInstance().registerFunction({ id: conditionPath }, async (input) => {
						const triggerInfo = {
							type: "stream",
							index
						};
						return trigger.condition?.(input, flowContext(this, triggerInfo, input));
					});
					triggerConfig.condition_function_id = conditionPath;
				}
				getInstance().registerTrigger({
					type: "stream",
					function_id,
					config: triggerConfig
				});
			}
		});
	}
	addStream(config, _streamPath) {
		this.streams[config.name] = new Stream(config);
	}
	initialize() {
		const hasJoin = Object.values(this.streams).some((stream$1) => stream$1.config.onJoin);
		const hasLeave = Object.values(this.streams).some((stream$1) => stream$1.config.onLeave);
		setupStepEndpoint(getInstance());
		if (this.authenticateStream) getInstance().registerFunction({ id: "motia::stream::authenticate" }, async (req) => {
			if (this.authenticateStream) {
				const context = flowContext(this, { type: "queue" });
				const input = {
					headers: req.headers,
					path: req.path,
					queryParams: req.query_params,
					addr: req.addr
				};
				return this.authenticateStream(input, context);
			}
		});
		if (hasJoin) {
			const function_id = "motia::stream::join";
			getInstance().registerFunction({ id: function_id }, async (req) => {
				const { stream_name, group_id, id, context: authContext } = req;
				const stream$1 = this.streams[stream_name];
				const context = flowContext(this, { type: "queue" });
				if (stream$1?.config.onJoin) return stream$1.config.onJoin({
					groupId: group_id,
					id
				}, context, authContext);
			});
			getInstance().registerTrigger({
				type: "stream:join",
				function_id,
				config: {}
			});
		}
		if (hasLeave) {
			const function_id = "motia::stream::leave";
			getInstance().registerFunction({ id: function_id }, async (req) => {
				const { stream_name, group_id, id, context: authContext } = req;
				const stream$1 = this.streams[stream_name];
				const context = flowContext(this, { type: "queue" });
				if (stream$1?.config.onLeave) await stream$1.config.onLeave({
					groupId: group_id,
					id
				}, context, authContext);
			});
			getInstance().registerTrigger({
				type: "stream:leave",
				function_id,
				config: {}
			});
		}
	}
};

//#endregion
//#region src/step.ts
function step(config, handler) {
	if (handler) return {
		config,
		handler
	};
	return {
		config,
		handle: (h) => ({
			config,
			handler: h
		})
	};
}

//#endregion
//#region src/triggers.ts
function http(method, path$1, options, condition) {
	return {
		type: "http",
		method,
		path: path$1,
		...options,
		condition
	};
}
/** @deprecated Use http() instead. Will be removed in a future version. */
function api(method, path$1, options, condition) {
	return http(method, path$1, options, condition);
}
function queue(topic, options, condition) {
	return {
		type: "queue",
		topic,
		...options,
		condition
	};
}
function cron(expression, condition) {
	return {
		type: "cron",
		expression,
		condition
	};
}
function state(condition) {
	return {
		type: "state",
		condition
	};
}
function stream(streamName, optionsOrCondition) {
	if (typeof optionsOrCondition === "function") return {
		type: "stream",
		streamName,
		condition: optionsOrCondition
	};
	const { groupId, itemId, condition } = optionsOrCondition ?? {};
	return {
		type: "stream",
		streamName,
		groupId,
		itemId,
		condition
	};
}

//#endregion
//#region src/types.ts
function jsonSchema(schema) {
	return z.toJSONSchema(schema, { target: "draft-7" });
}

//#endregion
export { Logger, Motia, StateManager, Stream, api, cron, generateStepId, getApiTriggers, getContext, getCronTriggers, getInstance, getQueueTriggers, http, initIII, isApiTrigger, isCronTrigger, isQueueTrigger, isStateTrigger, isStreamTrigger, jsonSchema, multiTriggerStep, queue, setupStepEndpoint, state, stateManager, step, stream };
