// iii-sdk/dist/telemetry.cjs
const require_otel_worker_gauges = require('./otel-worker-gauges-CLZyxI1P.cjs');
const require_index = require('./index.cjs');
let __opentelemetry_api_logs = require("@opentelemetry/api-logs");
let __opentelemetry_api = require("@opentelemetry/api");

exports.DEFAULT_BRIDGE_RECONNECTION_CONFIG = require_otel_worker_gauges.DEFAULT_BRIDGE_RECONNECTION_CONFIG;
exports.DEFAULT_INVOCATION_TIMEOUT_MS = require_otel_worker_gauges.DEFAULT_INVOCATION_TIMEOUT_MS;
exports.EngineFunctions = require_otel_worker_gauges.EngineFunctions;
exports.EngineTriggers = require_otel_worker_gauges.EngineTriggers;
exports.LogFunctions = require_otel_worker_gauges.LogFunctions;
exports.SeverityNumber = __opentelemetry_api_logs.SeverityNumber;
exports.SpanStatusCode = __opentelemetry_api.SpanStatusCode;
exports.WorkerMetricsCollector = require_otel_worker_gauges.WorkerMetricsCollector;
exports.currentSpanId = require_otel_worker_gauges.currentSpanId;
exports.currentTraceId = require_otel_worker_gauges.currentTraceId;
exports.extractBaggage = require_otel_worker_gauges.extractBaggage;
exports.extractContext = require_otel_worker_gauges.extractContext;
exports.extractTraceparent = require_otel_worker_gauges.extractTraceparent;
exports.getAllBaggage = require_otel_worker_gauges.getAllBaggage;
exports.getBaggageEntry = require_otel_worker_gauges.getBaggageEntry;
exports.getLogger = require_otel_worker_gauges.getLogger;
exports.getMeter = require_otel_worker_gauges.getMeter;
exports.getTracer = require_otel_worker_gauges.getTracer;
exports.initOtel = require_otel_worker_gauges.initOtel;
exports.injectBaggage = require_otel_worker_gauges.injectBaggage;
exports.injectTraceparent = require_otel_worker_gauges.injectTraceparent;
exports.registerWorkerGauges = require_otel_worker_gauges.registerWorkerGauges;
exports.removeBaggageEntry = require_otel_worker_gauges.removeBaggageEntry;
exports.safeStringify = require_otel_worker_gauges.safeStringify;
exports.setBaggageEntry = require_otel_worker_gauges.setBaggageEntry;
exports.shutdownOtel = require_otel_worker_gauges.shutdownOtel;
exports.stopWorkerGauges = require_otel_worker_gauges.stopWorkerGauges;
exports.withSpan = require_otel_worker_gauges.withSpan;
